rust/xaynet-server/src/settings.rs

```diff
@@ -558,6 +564,181 @@ where
+    /// ```text
+    /// XAYNET_S3__REGION="eu-west-1"
+    /// ```
+    ///
+    /// To connect to AWS-compatible services such as Minio, you need to specify a custom region.
+    ///
+    /// # Examples
+    ///
+    /// **TOML**
+    /// ```text
+    /// [s3]
+    /// region = ["minio", "http://localhost:8000"]
+    /// ```
+    ///
+    /// **Environment variable**
+    /// ```text
+    /// XAYNET_S3__REGION="minio http://localhost:8000"
+    /// ```
+    #[serde(deserialize_with = "deserialize_s3_region")]
+    pub region: Region,
+    #[validate]
+    #[serde(default)]
+    pub buckets: S3BucketsSettings,
+}
+
+#[derive(Debug, Validate, Deserialize)]
+pub struct S3BucketsSettings {
+    /// The bucket name in which the global models are stored.
+    /// Defaults to `global-models`.
+    ///
+    /// Please follow the [rules for bucket naming](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html)
+    /// when creating the name.
+    ///
+    /// # Examples
+    ///
+    /// **TOML**
+    /// ```text
+    /// [s3.buckets]
+    /// global_models = "global-models"
+    /// ```
+    ///
+    /// **Environment variable**
+    /// ```text
+    /// XAYNET_S3__BUCKETS__GLOBAL_MODELS="global-models"
+    /// ```
+    #[validate(custom = "validate_s3_bucket_name")]
+    global_models: String,
+}
+
+// Default value for the global models bucket
+impl Default for S3BucketsSettings {
+    fn default() -> Self {
+        Self {
+            global_models: String::from("global-models"),
+        }
+    }
+}
+
+// Validates the bucket name
+// [Rules for AWS bucket naming](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html)
+fn validate_s3_bucket_name(bucket_name: &str) -> Result<(), ValidationError> {
+    // https://stackoverflow.com/questions/50480924/regex-for-s3-bucket-name#comment104807676_58248645
+    // I had to use fancy_regex here because the std regex does not support `look-around`
+    let re =
+        Regex::new(r"(?!^(\d{1,3}\.){3}\d{1,3}$)(^[a-z0-9]([a-z0-9-]*(\.[a-z0-9])?)*$(?<!\-))")
+            .unwrap();
+    match re.is_match(bucket_name) {
+        Ok(true) => Ok(()),
+        Ok(false) => Err(ValidationError::new("invalid bucket name\n See here: https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html")),
+        // something went wrong with the regex engine
+        Err(_) => Err(ValidationError::new("can not validate bucket name")),
+    }
+}
+
+// A small wrapper to support the list type for environment variable values.
+// config-rs always converts a environment variable value to a string
+// https://github.com/mehcode/config-rs/blob/master/src/env.rs#L114 .
+// Strings however, are not supported by the deserializer of rusoto_core::Region (only sequences).
+// Therefore we use S3RegionVisitor to implement `visit_str` and thus support
+// the deserialization of rusoto_core::Region from strings.
+fn deserialize_s3_region<'de, D>(deserializer: D) -> Result<Region, D::Error>
```

Comment:
I have added some tests. It adds a lot of code tbh. Maybe we should spilt the settings.rs into multiple files in the near future. Unfortunately I have to use `serial` here because I set environment variables in some tests.
Let me know what you think🙂